### PR TITLE
Engine poc

### PR DESCRIPTION
**Pull request by shawn-hurley**

### CODE REVIEW

File: engine/conditions.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+type Condition struct {
+	// Optional When clause
+	When *Conditional `json:"when,omitempty"`
+
+	InnerCondition `json:"innerCondition,omitempty"`
+}
+
+type Conditional struct {
```
Comment (fabianvf) [inline engine/conditions.go]: Not sure I follow the separation between logical conditions and the innercondition, could or/and not be implemented as InnerConditions whose `Evaluate` method calls the `Evaluate` methods of the subconditions?

File: engine/engine.go [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+	Stop()
+}
+
+type ruleMessage struct {
+	rule       Rule
+	returnChan chan response
+}
+
+type response struct {
+	conditionResponse InnerConndtionResponse
```
Comment (fabianvf) [inline engine/engine.go]:
```suggestion
	conditionResponse InnerConditionResponse
```

File: engine/conditions.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
+package engine
+
+type InnerConndtionResponse struct {
```
Comment (fabianvf) [inline engine/conditions.go]:
```suggestion
type InnerConditionResponse struct {
```

File: engine/conditions.go [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+	Passed bool
+	// For each time the condition is hit, add all of the context.
+	// keys here, will be used in the message.
+	ConditionHitContext []map[string]string
+}
+
+type InnerCondition interface {
+	Evaluate() (InnerConndtionResponse, error)
```
Comment (fabianvf) [inline engine/conditions.go]:
```suggestion
	Evaluate() (InnerConditionResponse, error)
```

File: engine/engine.go [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+		fmt.Printf("done waiting for rules to be processed")
+	case <-ctx.Done():
+		fmt.Printf("Context canceled running of rules")
+	}
+	// Cannel running go-routine
+	cancelFunc()
+
+}
+
+func processRule(rule Rule) (InnerConndtionResponse, error) {
```
Comment (fabianvf) [inline engine/engine.go]:
```suggestion
func processRule(rule Rule) (InnerConditionResponse, error) {
```

File: engine/engine.go [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+
+	// Here is what a worker should run when getting a rule.
+	// For now, lets not fan out the running of conditions.
+
+	if rule.When.InnerCondition != nil {
+		//IF there is an inner conndtion here, that means there is only a single condition to run.
```
Comment (fabianvf) [inline engine/engine.go]:
```suggestion
		//IF there is an inner condition here, that means there is only a single condition to run.
```

File: engine/engine.go [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+		//Use output
+		return rule.When.InnerCondition.Evaluate()
+
+	} else {
+		if len(rule.When.And) > 0 {
+			return evaluateAndCondtions(rule.When.And)
+		} else if len(rule.When.Or) > 0 {
+			return evaluateOrConditions(rule.When.Or)
+		} else {
+			return InnerConndtionResponse{}, fmt.Errorf("invalid when condition")
```
Comment (fabianvf) [inline engine/engine.go]:
```suggestion
			return InnerConditionResponse{}, fmt.Errorf("invalid when condition")
```

File: engine/engine.go [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+		}
+	}
+}
+
+func evaluateAndCondtions(conditions []Condition) (InnerConndtionResponse, error) {
```
Comment (fabianvf) [inline engine/engine.go]:
```suggestion
func evaluateAndCondtions(conditions []Condition) (InnerConditionResponse, error) {
```

File: engine/engine.go [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+	// Make sure we allow for short circt.
+
+	if len(conditions) == 0 {
+		return InnerConndtionResponse{}, fmt.Errorf("condtions must not be empty while evaluationg")
```
Comment (fabianvf) [inline engine/engine.go]:
```suggestion
		return InnerConditionResponse{}, fmt.Errorf("condtions must not be empty while evaluationg")
```

File: engine/engine.go [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+	}
+
+	fullResponse := InnerConndtionResponse{Passed: true}
```
Comment (fabianvf) [inline engine/engine.go]:
```suggestion
	fullResponse := InnerConditionResponse{Passed: true}
```

File: engine/engine.go [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+	for _, c := range conditions {
+		if c.When != nil {
+			var response InnerConndtionResponse
```
Comment (fabianvf) [inline engine/engine.go]:
```suggestion
			var response InnerConditionResponse
```

File: engine/engine.go [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+				return response, err
+			}
+			fullResponse.ConditionHitContext = append(fullResponse.ConditionHitContext, response.ConditionHitContext...)
+		}
+	}
+
+	return fullResponse, nil
+}
+
+func evaluateOrConditions(conditions []Condition) (InnerConndtionResponse, error) {
```
Comment (fabianvf) [inline engine/engine.go]:
```suggestion
func evaluateOrConditions(conditions []Condition) (InnerConditionResponse, error) {
```

File: engine/engine.go [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+	for _, c := range conditions {
+		if c.When != nil {
+			var response InnerConndtionResponse
```
Comment (fabianvf) [inline engine/engine.go]:
```suggestion
			var response InnerConditionResponse
```

File: engine/engine.go [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+			if response.Passed {
+				return response, err
+			}
+		} else {
+			response, err := c.InnerCondition.Evaluate()
+			if response.Passed {
+				return response, err
+			}
+			if err != nil {
+				return InnerConndtionResponse{}, err
```
Comment (fabianvf) [inline engine/engine.go]:
```suggestion
				return InnerConditionResponse{}, err
```

File: engine/engine.go [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+			}
+		}
+	}
+
+	return InnerConndtionResponse{}, nil
```
Comment (fabianvf) [inline engine/engine.go]:
```suggestion
	return InnerConditionResponse{}, nil
```

File: engine/engine_test.go [hunk 16]
```diff
@@ -0,0 +1,293 @@
...
+	"time"
+)
+
+type testInerCondition struct {
+	err   error
+	ret   bool
+	sleep bool
+}
+
+func (t testInerCondition) Evaluate() (InnerConndtionResponse, error) {
```
Comment (fabianvf) [inline engine/engine_test.go]:
```suggestion
func (t testInnerCondition) Evaluate() (InnerConditionResponse, error) {
```

File: engine/engine_test.go [hunk 17]
```diff
@@ -0,0 +1,293 @@
+package engine
+
+import (
+	"context"
+	"testing"
+	"time"
+)
+
+type testInerCondition struct {
```
Comment (fabianvf) [inline engine/engine_test.go]:
```suggestion
type testInnerCondition struct {
```

File: engine/engine_test.go [hunk 18]
```diff
@@ -0,0 +1,293 @@
...
+	err   error
+	ret   bool
+	sleep bool
+}
+
+func (t testInerCondition) Evaluate() (InnerConndtionResponse, error) {
+	if t.sleep {
+		time.Sleep(5 * time.Second)
+	}
+	return InnerConndtionResponse{Passed: t.ret}, t.err
```
Comment (fabianvf) [inline engine/engine_test.go]:
```suggestion
	return InnerConditionResponse{Passed: t.ret}, t.err
```

File: engine/engine_test.go [hunk 19]
```diff
@@ -0,0 +1,293 @@
...
+}
+
+func createTestInnerCondition(b bool, e error, sleep bool) InnerCondition {
+	return &testInerCondition{
```
Comment (fabianvf) [inline engine/engine_test.go]:
```suggestion
	return &testInnerCondition{
```

File: provider/provider.go [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+	Evaluate(cap string, conditionInfo interface{}) (lib.ProviderEvaluateResponse, error)
+}
+
+type ProviderCondition struct {
+	Client
+	Capability    string
+	ConditionInfo interface{}
+}
+
+func (p *ProviderCondition) Evaluate() (engine.InnerConndtionResponse, error) {
```
Comment (fabianvf) [inline provider/provider.go]:
```suggestion
func (p *ProviderCondition) Evaluate() (engine.InnerConditionResponse, error) {
```

File: provider/provider.go [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+	resp, err := p.Client.Evaluate(p.Capability, p.ConditionInfo)
+	if err != nil {
+		// If an error always just return the empty
+		return engine.InnerConndtionResponse{}, err
```
Comment (fabianvf) [inline provider/provider.go]:
```suggestion
		return engine.InnerConditionResponse{}, err
```

File: provider/provider.go [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+	}
+
+	return engine.InnerConndtionResponse{
```
Comment (fabianvf) [inline provider/provider.go]:
```suggestion
	return engine.InnerConditionResponse{
```

File: engine/engine.go [hunk 23]
```diff
@@ -0,0 +1,218 @@
...
+}
+
+type ruleEngine struct {
+	// Buffereed channel where Rule Processors are watching
+	ruleProcessing chan ruleMessage
+	cancelFunc     context.CancelFunc
+}
+
+func CreateRuleEngine(ctx context.Context, workers int) RuleEngine {
+	// Only allow for 10 rules to be processed at once.
```
Comment (fabianvf) [inline engine/engine.go]: why?

File: engine/engine.go [hunk 24]
```diff
@@ -0,0 +1,218 @@
...
+			select {
+			case response := <-ret:
+				if response.conditionResponse.Passed {
+					responses = append(responses, response)
+
+				} else {
+					// Log that rule did not pass
+					fmt.Printf("Rule did not pass")
+				}
+				responses = append(responses, response)
```
Comment (fabianvf) [inline engine/engine.go:99]: What happens to these responses?

File: engine/engine_test.go [hunk 25]
```diff
@@ -0,0 +1,293 @@
...
+
+func TestEvaluateAndCondtions(t *testing.T) {
+
+	testCases := []struct {
+		Name       string
+		Conditions []Condition
+		IsError    bool
+		IsPassed   bool
+	}{
+		// {
```
Comment (fabianvf) [inline engine/engine_test.go]: Why are these disabled?

File: examples/java/.metadata/.log [hunk 26]
```diff
@@ -0,0 +1,36 @@
+!SESSION 2022-08-22 11:34:46.747 -----------------------------------------------
```
Comment (fabianvf) [inline examples/java/.metadata/.log]: uncommit this I think

File: examples/java/.metadata/.plugins/org.eclipse.core.resources/.root/.indexes/properties.version [hunk 27]
```diff
@@ -0,0 +1 @@
+
```
Comment (fabianvf) [inline examples/java/.metadata/.plugins/org.eclipse.core.resources/.root/.indexes/properties.version]: probably should uncommit a lot of these files and add them to .gitignore

File: main.go [hunk 28]
```diff
@@ -3,116 +3,59 @@ package main
...
-				Import: "pkg/apis/apiextensions/v1beta1.CustomResourceDefinition",
-				// TODO(shawn-hurley) - copy the windup ability to intersparse known text here.
-				Message: "Use of deprecated and removed API",
-			},
-		},
-	},
-}
+const (
+	// This must eventually be a default that makes sense, and overrideable by env var or flag.
+	SETTING_FILE_PATH = "/home/shurley/repos/jsonrpc-golang/provider_settings.json"
```
Comment (fabianvf) [inline main.go]: probably should make this configurable pre-merge

File: main.go [hunk 29]
```diff
@@ -3,116 +3,59 @@ package main
...
+		providers[config.Name] = provider
 	}
-	fmt.Printf("connection initialized: %#v", result.Capabilities.WorkspaceSymbolProvider)
 
-	for _, r := range workspaceRules {
-		if r.GoImportRule != nil {
-			if loc, found := findReferences(ctx, rpc, absolutePath, r.GoImportRule.Import); found {
-				fmt.Printf("\n%v\nlocation: %v:%v", r.GoImportRule.Message, loc.URI, loc.Range.Start.Line)
-			}
+	// Now that we have all the providers, we need to start them.
```
Comment (fabianvf) [inline main.go]: Do we want to start them now, or start them once we know which providers are actually referenced in the rules?

File: parser/rule_parser.go [hunk 30]
```diff
@@ -0,0 +1,206 @@
...
+
+	"github.com/shawn-hurley/jsonrpc-golang/engine"
+	"github.com/shawn-hurley/jsonrpc-golang/provider"
+)
+
+type RuleParser struct {
+	ProviderNameToClient map[string]provider.Client
+}
+
+// This will
```
Comment (fabianvf) [inline parser/rule_parser.go]: will what

File: parser/rule_parser.go [hunk 31]
```diff
@@ -0,0 +1,206 @@
...
+	ruleMap := []map[string]interface{}{}
+
+	err = json.Unmarshal(content, &ruleMap)
+	if err != nil {
+		return nil, err
+	}
+
+	rules := []engine.Rule{}
+	for _, ruleMap := range ruleMap {
+		// Rule's right now only contain two top level things, message and when.
```
Comment (fabianvf) [inline parser/rule_parser.go]:
```suggestion
		// Rules right now only contain two top level things, message and when.
```

File: parser/rule_parser.go [hunk 32]
```diff
@@ -0,0 +1,206 @@
...
+	"github.com/shawn-hurley/jsonrpc-golang/engine"
+	"github.com/shawn-hurley/jsonrpc-golang/provider"
+)
+
+type RuleParser struct {
+	ProviderNameToClient map[string]provider.Client
+}
+
+// This will
+func (r *RuleParser) LoadRules(filepath string) ([]engine.Rule, error) {
```
Comment (fabianvf) [inline parser/rule_parser.go]: could also return a list of used providers so we know what we need to start?

File: parser/rule_parser.go [hunk 33]
```diff
@@ -0,0 +1,206 @@
...
+	err = json.Unmarshal(content, &ruleMap)
+	if err != nil {
+		return nil, err
+	}
+
+	rules := []engine.Rule{}
+	for _, ruleMap := range ruleMap {
+		// Rule's right now only contain two top level things, message and when.
+		// When is where we need to handle conditions
+		message, ok := ruleMap["message"].(string)
```
Comment (fabianvf) [inline parser/rule_parser.go:37]: Assume this is a placeholder for now as we'll likely have more structured actions at some point, probably worth a TODO if so

File: provider/golang/provider.go [hunk 34]
```diff
@@ -0,0 +1,157 @@
...
+func (p *golangProvider) Capabilities() ([]string, error) {
+	return []string{
+		"referenced",
+	}, nil
+}
+
+func (p *golangProvider) Evaluate(cap string, conditionInfo interface{}) (lib.ProviderEvaluateResponse, error) {
+
+	// query := conditionInfo["referenced"].(string)
+	query := "query"
```
Comment (fabianvf) [inline provider/golang/provider.go]: assume we want the line above this

File: provider/java/provider.go [hunk 35]
```diff
@@ -0,0 +1,161 @@
...
+const BUNDLES_INIT_OPTION = "bundles"
+const WORKSPACE_INIT_OPTION = "workspace"
+
+func NewJavaProvider(config lib.Config) *javaProvider {
+
+	// Get the provider config out for this config.
+
+	// Getting values out of provider config
+
+	// Get
```
Comment (fabianvf) [inline provider/java/provider.go]: unfinished comment

File: provider/java/provider.go [hunk 36]
```diff
@@ -0,0 +1,161 @@
...
+		"referenced",
+	}, nil
+}
+
+func (p *javaProvider) Evaluate(cap string, conditionInfo interface{}) (lib.ProviderEvaluateResponse, error) {
+	return lib.ProviderEvaluateResponse{}, nil
+}
+func (p *javaProvider) Init(ctx context.Context) error {
+
+	cmd := exec.CommandContext(ctx, "/home/shurley/repos/eclipse.jdt.ls/org.eclipse.jdt.ls.product/target/repository/bin/jdtls",
```
Comment (fabianvf) [inline provider/java/provider.go]: need to fix this hardcoded path

Comment (fabianvf) [reply provider/java/provider.go]: probably needs to be part of the provider config?

File: provider/java/provider.go [hunk 37]
```diff
@@ -0,0 +1,161 @@
...
+		"-configuration",
+		"/home/shurley/config",
+		"-data",
+		p.workspace,
+	)
+	stdin, err := cmd.StdinPipe()
+	if err != nil {
+		fmt.Printf("HERE!!!!!! - %v", err)
```
Comment (fabianvf) [inline provider/java/provider.go]: remove these prints (or make the messages useful)

File: provider/java/provider.go [hunk 38]
```diff
@@ -0,0 +1,161 @@
...
+		}
+	}()
+
+	p.rpc = rpc
+	p.ctx = ctx
+	p.initialization(ctx)
+	return nil
+}
+
+// This can probably be shared between the two
```
Comment (fabianvf) [inline provider/java/provider.go]: which two?

File: provider/java/provider.go [hunk 39]
```diff
@@ -0,0 +1,161 @@
...
+		break
+	}
+	if err := p.rpc.Notify(ctx, "initialized", &protocol.InitializedParams{}); err != nil {
+		fmt.Printf("initialized failed: %v", err)
+	}
+	fmt.Printf("\njava connection initialized: %#v", result.Capabilities.WorkspaceSymbolProvider)
+}
+
+func (p *javaProvider) GetAllSymbols(query string) []protocol.WorkspaceSymbol {
+	//workspace/executeCommand '{"command": "io.konveyor.tackle.ruleEntry", "arguments": {"query":"*customresourcedefinition","project": "java"}}'
```
Comment (fabianvf) [inline provider/java/provider.go]: comment needs some elaboration

File: provider/java/provider.go (hunk 36 above)
Comment (fabianvf) [reply provider/java/provider.go]: looks like it already is, just need to use the value

File: rules_bak/types.go [hunk 40]
```diff
@@ -1,4 +1,4 @@
-package rules
```
Comment (fabianvf) [inline rules_bak/types.go]: remove this?

**Review by fabianvf — CHANGES_REQUESTED** [review]

File: engine/engine.go (hunk 23 above)
Comment (shawn-hurley) [reply engine/engine.go]: This should be changed to only allow for 10 rules to being waiting at once

File: engine/engine.go (hunk 24 above)
Comment (shawn-hurley) [reply engine/engine.go:99]: for now, nothing, we don't have any output format for this

File: engine/engine_test.go (hunk 25 above)
Comment (shawn-hurley) [reply engine/engine_test.go]: 🤦🏻‍♂️

File: examples/java/.metadata/.log (hunk 26 above)
Comment (shawn-hurley) [reply examples/java/.metadata/.log]: think I want todelete this whole thing

Comment (shawn-hurley) [reply examples/java/.metadata/.log]: and add to git ignore like you mentioned

File: main.go (hunk 28 above)
Comment (shawn-hurley) [reply main.go]: probably make it a relative path in the constants and then can add cobra and such as a followup? thoughts?

File: parser/rule_parser.go (hunk 30 above)
Comment (shawn-hurley) [reply parser/rule_parser.go]: 🤦🏻‍♂️

File: engine/engine.go (hunk 24 above)
Comment (shawn-hurley) [reply engine/engine.go:99]: adding todo at line 126 for this

File: engine/conditions.go [hunk 41]
```diff
@@ -0,0 +1,66 @@
+package engine
+
+import "fmt"
+
+type ConndtionResponse struct {
```
Comment (fabianvf) [inline engine/conditions.go]: hmm looks like the commit that fixed these was overwritten,  `ConndtionResponse` => `ConditionResponse`

File: engine/conditions.go [hunk 42]
```diff
@@ -0,0 +1,66 @@
...
+	return fullResponse, nil
+}
+
+type OrCondition struct {
+	Conditions []Conditional
+}
+
+func (o OrCondition) Evaluate() (ConndtionResponse, error) {
+	if len(o.Conditions) == 0 {
+		return ConndtionResponse{}, fmt.Errorf("condtions must not be empty while evaluationg")
```
Comment (fabianvf) [inline engine/conditions.go]: `condtions` => `conditions` in the string

File: engine/conditions.go [hunk 43]
```diff
@@ -0,0 +1,66 @@
...
+	}
+
+	for _, c := range o.Conditions {
+		response, err := c.Evaluate()
+		// Short cirtcut loop if one and condition passes we can move on
```
Comment (fabianvf) [inline engine/conditions.go]: s/and/or/

File: engine/engine.go [hunk 44]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+type response struct {
+	conditionResponse ConndtionResponse
+	err               error
+	rule              Rule
+}
+
+type ruleEngine struct {
+	// Buffereed channel where Rule Processors are watching
```
Comment (fabianvf) [inline engine/engine.go]: `Buffereed` => `Buffered`

File: examples/java/.project [hunk 45]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (fabianvf) [inline examples/java/.project]: can this be deleted as well?

File: examples/java/.classpath [hunk 46]
```diff
@@ -0,0 +1,44 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (fabianvf) [inline examples/java/.classpath]: Can this be deleted?

File: examples/java/.settings/org.eclipse.core.resources.prefs [hunk 47]
```diff
@@ -0,0 +1,4 @@
+eclipse.preferences.version=1
```
Comment (fabianvf) [inline examples/java/.settings/org.eclipse.core.resources.prefs]: Can the .settings directory be removed + ignored?

File: parser/rule_parser.go (hunk 30 above)
Comment (fabianvf) [reply parser/rule_parser.go]: will what tho

File: provider/golang/provider.go (hunk 34 above)
Comment (fabianvf) [reply provider/golang/provider.go]: bump

**Review by fabianvf — CHANGES_REQUESTED** [review]

File: provider/golang/provider.go (hunk 34 above)
Comment (shawn-hurley) [reply provider/golang/provider.go]: fixed for real this time

File: parser/rule_parser.go (hunk 30 above)
Comment (shawn-hurley) [reply parser/rule_parser.go]: updated for real this time

**Pull request merged**